File: src/Longest Subsequence with Adjacent Diff as 1/Code/solution.js
/**
 * @param {number[]} arr
 * @returns {number}
 */

class Solution {
  longestSubseq(arr) {
    // dp stores the longest valid subsequence ending with each value.
    const dp = new Map();

    // This stores the longest subsequence found so far.
    let ans = 0;

    // Process elements from left to right so subsequence order is preserved.
    for (const x of arr) {
      // Get the best subsequence ending at x - 1, or 0 if it does not exist.
      const left = dp.get(x - 1) || 0;

      // Get the best subsequence ending at x + 1, or 0 if it does not exist.
      const right = dp.get(x + 1) || 0;

      // Extend the better neighboring subsequence using the current value x.
      const current = Math.max(left, right) + 1;

      // Store the longest subsequence ending with x.
      dp.set(x, current);

      // Update the overall maximum answer.
      ans = Math.max(ans, current);
    }

    // Return the length of the longest valid subsequence.
    return ans;
  }
}

File: src/Longest Subsequence with Adjacent Diff as 1/Code/solution.py
class Solution:
    def longestSubseq(self, arr):
        # dp[x] stores the longest valid subsequence ending with value x.
        dp = {}

        # This stores the longest subsequence found so far.
        ans = 0

        # Process elements from left to right to maintain subsequence order.
        for x in arr:
            # Get the best subsequence ending at x - 1, or 0 if it does not exist.
            left = dp.get(x - 1, 0)

            # Get the best subsequence ending at x + 1, or 0 if it does not exist.
            right = dp.get(x + 1, 0)

            # Extend the better neighboring subsequence with the current value x.
            current = max(left, right) + 1

            # Store the longest subsequence ending with x.
            dp[x] = current

            # Update the overall maximum answer.
            ans = max(ans, current)

        # Return the length of the longest valid subsequence.
        return ans